### PR TITLE
Split tests part 2

### PR DESCRIPTION
**Pull request by antrix1989**
Changed isEqual method for both ADTokenCacheItem & ADUserInformation.
Some questions that I have here:
- Can we allow to use -init of ADUserInformation?
- Do we need to enable "DISABLEDtestIsSameUser" test?
- Can we remove - (BOOL)isSameUser:(ADTokenCacheItem *)other from ADTokenCacheItem?

### CODE REVIEW

File: ADAL/src/cache/ADTokenCacheItem.m [hunk 1]
```diff
@@ -121,7 +121,8 @@ - (BOOL)isSameUser:(ADTokenCacheItem*) other
     
     if ([self isEmptyUser])
         return [other isEmptyUser];
-    return (nil != other.userInformation && [_userInformation.userId isEqualToString:other.userInformation.userId]);
+    
+    return [self.userInformation isEqual:other.userInformation];
```
Comment (oldalton) [inline ADAL/src/cache/ADTokenCacheItem.m]: nit: please update comment for this method, currently:

`/*! Verifies if the user (as defined by userId) is the same between the two items. */`

That is not exactly true anymore, since now it will verify all the properties and not only userId.

File: ADAL/tests/unit/ADLoggerTests.m [hunk 2]
```diff
@@ -35,20 +35,47 @@ - (void)setUp
...
+    [ADLogger setNSLogging:NO];
+    
+    XCTAssertFalse([ADLogger getNSLogging]);
 }
 
-- (void)testMessageNoThrowing
+#pragma mark - log:context:message:errorCode:info:correlationId:userInfo
+
+- (void)testLog_whenLogLevelNoLogContextNilMessageValidErrorCodeSucceededInfoValidCorrelationIdNilUserInfoNil_shouldNotThrow
 {
```
Comment (oldalton) [inline ADAL/tests/unit/ADLoggerTests.m]: nit: Those 3 names are a little bit unreadable...

File: ADAL/tests/unit/ADTokenCacheItemTests.m [hunk 3]
```diff
@@ -43,110 +43,480 @@ - (void)tearDown
...
+    ADTokenCacheItem *item2 = [ADTokenCacheItem new];
+    item2.userInformation = nil;
+    
+    XCTAssertTrue([item1 isSameUser:item2]);
+    XCTAssertTrue([item2 isSameUser:item1]);
+}
+
+// TODO: do we need this behaviour?
+- (void)DISABLEDtestIsSameUser_whenOtherUserIdStartsWithSpaceButOtherPartIsIdenticalToCurrent_shouldReturnTrue
+{
```
Comment (oldalton) [inline ADAL/tests/unit/ADTokenCacheItemTests.m]: why this is disabled?

File: ADAL/tests/unit/ADTokenCacheItemTests.m [hunk 4]
```diff
@@ -43,110 +43,480 @@ - (void)tearDown
...
+    ADTokenCacheItem *lhs = [ADTokenCacheItem new];
+    [lhs setValue:@{@"k1":@"v1"} forKey:@"additionalServer"];
+    ADTokenCacheItem *rhs = [ADTokenCacheItem new];
+    [rhs setValue:@{@"k2":@"v2"} forKey:@"additionalServer"];
+    
+    XCTAssertNotEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAdditionalServerIsEqual_shouldReturnTrue
```
Comment (oldalton) [inline ADAL/tests/unit/ADTokenCacheItemTests.m]: There's a test for each property to test when it's equal or not equal, but I couldn't find any tests for when some of the properties are nil. On the other hand, I'm feeling a little bit bad about creating a separate test for each property for each condition (e.g testing for nil would now add 12 more tests....)

File: ADAL/tests/unit/ADUserInformationTests.m [hunk 5]
```diff
@@ -41,14 +41,141 @@ - (void)tearDown
...
+{
+    ADUserInformation *lhs = [self adCreateUserInformation:@"eric_cartman@contoso.com"];
+    ADUserInformation *rhs = [self adCreateUserInformation:@"eric_cartman@contoso.com"];
+    
+    XCTAssertEqualObjects(lhs, rhs);
+}
+
+- (void)testIsEqual_whenAllClaimsIsNotEqual_shouldReturnFalse
+{
+    ADUserInformation *lhs = [self createEmptyUserInformation];
```
Comment (oldalton) [inline ADAL/tests/unit/ADUserInformationTests.m]: Same thing here, some of the ADUserInformation properties might be nil and I would like that to be tested to at least some extent.

File: ADAL/tests/unit/ADUserInformationTests.m [hunk 6]
```diff
@@ -41,14 +41,141 @@ - (void)tearDown
...
+
 - (void) testCopy
 {
     ADUserInformation* userInfo = [self adCreateUserInformation:@"eric_cartman@contoso.com"];
     XCTAssertNotNil(userInfo);
     
     ADUserInformation* copy = [userInfo copy];
     XCTAssertNotNil(copy);
-    XCTAssertNotEqualObjects(copy, userInfo);
+    XCTAssertEqualObjects(copy, userInfo);
```
Comment (oldalton) [inline ADAL/tests/unit/ADUserInformationTests.m]: Do we still need to separately check userId, givenName and familyName now after your changes?

File: ADAL/src/cache/ADTokenCacheItem.m (hunk 1 above)
Comment (antrix1989) [inline ADAL/src/cache/ADTokenCacheItem.m]: sure, I can update the comment, but can we remove completely 
`- (BOOL)isSameUser:(ADTokenCacheItem *)other`
method? I've checked and it is not used in ADAL anymore, but it leaves in public headers, so it might affect other users of the library. @RPangrle @oldalton what do you guys think, can we remove it?

**Comment by RPangrle** [comment]
To answer your questions:

- Can we allow to use -init of ADUserInformation?

I generally don't see the harm in this, even though it probably won't allow the developer to accomplish much, which is why I generally don't see the harm. ADUserInformation *should* be mostly a wrapper around a JSON blob, so init'ing one without it shouldn't be terribly useful.

- Do we need to enable "DISABLEDtestIsSameUser" test?

Maybe?

- Can we remove - (BOOL)isSameUser:(ADTokenCacheItem *)other from ADTokenCacheItem?

Assuming it's not being used anywhere in product code, yes.

File: ADAL/src/cache/ADTokenCacheItem.m (hunk 1 above)
Comment (oldalton) [inline ADAL/src/cache/ADTokenCacheItem.m]: I'm fine removing it, make sure you remove the corresponding tests as well

File: ADAL/tests/unit/ADLoggerTests.m (hunk 2 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADLoggerTests.m]: simplified

Comment (oldalton) [inline ADAL/tests/unit/ADLoggerTests.m]: Thanks a lot 👍

File: ADAL/tests/unit/ADTokenCacheItemTests.m (hunk 3 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADTokenCacheItemTests.m]: it was disabled because "isEqual" logic is changed and that test was failing. UserId was passing inside "allClaims" dictionary without trimming, and as a result 2 UserInfo objects were are not equal. Since we decided to remove "IsSameUser" method, that test is removed as well. Testing of UserInfo equality is moved to ADUserInformationTests.

File: ADAL/tests/unit/ADTokenCacheItemTests.m (hunk 4 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADTokenCacheItemTests.m]: There are 2 types of tests:
- Verify that objects are equal when all properties are set:
`testIsEqual_whenAllPropertiesAreEqual_shouldReturnTrue`
- Verify that objects are equal when all properties are nil and only one property is set, for example:
```
testIsEqual_whenResourceIsEqual_shouldReturnTrue
testIsEqual_whenResourceIsNotEqual_shouldReturnFalse
```
We can add other tests that verify mix of nil and not nil properties, but I think those test will overlap with current ones.

File: ADAL/src/ADUserInformation.m [hunk 7]
```diff
@@ -259,6 +259,35 @@ - (id)copyWithZone:(NSZone *)zone
...
+    if (![object isKindOfClass:[ADUserInformation class]])
+    {
+        return NO;
+    }
+    
+    ADUserInformation *rhs = (ADUserInformation *)object;
+    
+    BOOL result = YES;
+    result &= [self.allClaims isEqualToDictionary:rhs.allClaims] || (self.allClaims == rhs.allClaims);
+    result &= [self.rawIdToken isEqual:rhs.rawIdToken] || (self.rawIdToken == rhs.rawIdToken);
```
Comment (antrix1989) [inline ADAL/src/ADUserInformation.m]: For this particular case yes, but lets think in another way: if we decide to inherit it and create ADMutableUserInformation, where user can change properties, then "isEqual" will not work properly. I think we should not "assume" that other properties are parsed from rawIdToken token here.

File: ADAL/tests/unit/ADUserInformationTests.m (hunk 6 above)
Comment (antrix1989) [inline ADAL/tests/unit/ADUserInformationTests.m]: We still need to check them in order to be 100% sure that copy method works properly. Otherwise if we just compare 2 instances of userInfo and our isEqual method is broken this test might give you false positive.

File: ADAL/src/ADUserInformation.m (hunk 7 above)
Comment (antrix1989) [inline ADAL/src/ADUserInformation.m]: Changed, as it is not so important. Either way should work for now.

File: ADAL/tests/unit/ADTokenCacheItemTests.m (hunk 4 above)
Comment (oldalton) [inline ADAL/tests/unit/ADTokenCacheItemTests.m]: I'm fine like that, just wanted to make sure we test different conditions
